README.md
# zxdb for VS Code

Note: This extension is merged into the [Fuchsia
extension](https://marketplace.visualstudio.com/items?itemName=fuchsia-authors.vscode-fuchsia) and
will be deprecated shortly. Please use the Fuchsia extension for debugger support.

[The VS Code zxdb extension](https://marketplace.visualstudio.com/items?itemName=fuchsia-authors.zxdb)
provides IDE based [zxdb](https://fuchsia.dev/fuchsia-src/development/debugger?hl=en) debugger
support for developers working with the open source [Fuchsia](https://fuchsia.dev/) operating system.

Note: This is not an officially supported Google product

## Features

- One click debugging - Press **F5** to start debugging on Fuchsia devices.
- Launch or Attach to a process from VS Code.
- zxdb console is parallely accessible in the integrated terminal.
- Works with VS Code Remote.
- Breakpoints, Exceptions, Stacktraces, Local Variables, Registers, Debug Hover
- Pause/Resume, Step In/Step Out/Step Over, Restart

![This gif shows a demo of the features.](images/zxdb-demo.gif)

## Prerequisites

1. If you don’t have the fuchsia source tree, see the
   [Fuchsia setup guide](https://fuchsia.dev/fuchsia-src/get-started/get_fuchsia_source?hl=en).
1. If this is the first time using zxdb, see the
   [zxdb setup guide](https://fuchsia.dev/fuchsia-src/development/debugger/running?hl=en).
1. If this is the first time using VS Code, see
   [VS Code setup for Fuchsia](https://fuchsia.dev/fuchsia-src/development/editors/vscode?hl=en).

## Quick start

- **Step 0:** Open VS Code application and install **zxdb** extension from marketplace
  (Click **View** then **Extensions** and search for **zxdb**, select **zxdb** and click on
  **install**).

- **Step 1:** Make sure that the target device is running fuchsia and the package server
  (like `fx serve`) has started on the development host. Make sure to have installed the
  [zxdb](https://marketplace.visualstudio.com/items?itemName=fuchsia-authors.zxdb)
  extension from VS Code marketplace.

- **Step 2:** Open VS Code application. Open the Fuchsia source directory (Click **File** then
  **Open..** or **Open Folder..** then select the Fuchsia directory).

- **Step 3:** Open the source code of the program you want to debug and add breakpoints by clicking
  on the editor margin or using **F9** on the current line.

  ![This figure show how to add breakpoints in VS Code.](images/breakpoint.png)

- **Step 4:** Add a launch configuration (Click **Run** then **Add Configuration**, if there is a
  drop down select zxdb). Update fields marked as _TODO_ in the launch.json. You only need to
  do this step whenever you want to update the configuration. See [this](#launch-configuration)
  section for more launch options.

  ![This figure shows a sample launch configuration.](images/test-launch-configuration.png)

- **Step 5:** Start debugging (press **F5** or **Play** button). See this
  [guide](https://code.visualstudio.com/docs/editor/debugging) to learn about VS Code debugger.

![This figure shows the different parts of the debugger view in VS Code.](images/vscode-overview.png)

See [troubleshooting guide](https://github.com/google/vscode-zxdb/blob/main/TROUBLESHOOTING.md)
if you face any issues.

## Launch Configuration

VS Code keeps debugging configuration information in a `launch.json` file located in a .vscode
folder in your workspace (project root folder) or in your
[user settings](https://code.visualstudio.com/docs/editor/debugging#_global-launch-configuration) or
[workspace settings](https://code.visualstudio.com/docs/editor/multi-root-workspaces#_workspace-launch-configurations).
In Step 4 of [Quick Start](#quick-start), this file is created automatically if it doesn’t
already exist.

You can add as many launch configurations as you need and pick one before starting to debug.

Follow these steps to update your launch configuration:

1. Click **Run** then **Open Configuration** to open launch configuration.
1. To add new configurations, Click **Run** then **Add Configuration** and scroll down to zxdb
   configuration snippets (_Hint: type zxdb to get there faster_).

   ![This image shows different launch configuration options.](images/launch-configurations.png)

1. Pick a configuration that matches your workflow.

   1. **Launch** configurations: Use this if you want to use VS Code to both launch and
      debug the program.

      Required fields:

      - Launch command: Shell command to launch the program.

        Example launch commands: fx test -o hello-world-test, fx shell i2cutil, fx run hello-world

      - Process name: Name of the process that will be started by the launch command.

        Example process name: hello-world-test, hello-world (partial names are also valid)

   1. **Attach** configurations: Use this if you want to use VS Code only for debugging.
      You need to launch the program manually.

      Required fields:

      - Process name: Name of the process to debug.

   An example launch configuration file can be found [here](https://github.com/google/vscode-zxdb/blob/main/example-launch.json)

1. Save the `launch.json`. The configuration names would appear in the launch configuration dropdown.

   ![This image shows the launch configuration dropdown.](images/launch-dropdown.png)

1. Select the required configuration from the dropdown and start debugging.

> **Tip**: Launch configuration can be quickly accessed by clicking on the gear button as shown below -
>
> ![This image shows the gear button that can be used to open launch configuration.](images/open-launch-configuration.png)

> **Note**: _Launch(zxdb) Input box_ / _Attach(zxdb) Input box_ configurations allow you to enter the
> customizable fields into a pop-up input box during every run just before the debugger starts.
> Use this as a generic configuration. However, inputs will not be saved and need to be entered in
> every run, which could be a compelling reason to update `launch.json` manually as described above.
>
> ![This image shows input box that takes launch configuration fields.](images/input-box.png)

## Extension configuration

To configure the extension, open settings (`Ctrl/Cmd + Shift + P` and type
**Preferences: Open Settings(UI)**). Type zxdb in the search bar to view and update the zxdb
extension configurations. After changing any of these configurations reload the VS Code window
(`Ctrl/Cmd + Shift + P` and type **Reload Window**) for it to take effect.

![This image shows different extension settings for zxdb.](images/extension-settings.png)

## Report issues

Please report issues/feedback
[here](https://bugs.fuchsia.dev/p/fuchsia/issues/entry?components=Tools%3Ezxdb;cc=puneetha@google.com;).

Please collect logs by following these steps to append to the report:

- For verbose logs, set `zxdb.debug: true` as described in the [this](#extension-configuration)
  section.
- Go to the output panel (Click **View** then **Output**)
- From the dropdown in the top right of that panel select zxdb.
- Copy all the logs (Click anywhere in the output panel, then press `Ctrl/Cmd + A` followed by
  `Ctrl/Cmd + C`). Paste the logs to a text file and attach it to the bug report.

## Release Notes

See the [CHANGELOG](https://github.com/google/vscode-zxdb/blob/main/CHANGELOG.md)


CHANGELOG.md
# Change Log

All notable changes to the "zxdb" extension will be documented in this file.

## [0.0.1]

- Initial release : Extension for fuchsia debugger
  - Visual studio code extension for zxdb - the fuchsia debugger.

## [0.0.2]

- Added logo

## [1.0.0]

- Marketplace release

## [1.0.1]

- Trim process name to 32 characters. Update Readme.

## [1.0.2]

- Use "ffx debug" instead of "fx debug" which supports out-of-tree debugging.
- Update the dependency.

## [1.0.3]

- Trim process name to 31 characters because NUL is the last character in
  Zircon.

## [1.0.4]

- Add deprecation warning. Please use [Fuchsia extension](https://marketplace.visualstudio.com/items?itemName=fuchsia-authors.vscode-fuchsia) instead.


src/test/runTest.ts
// Copyright 2021 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     https://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import * as path from 'path';
import {runTests} from 'vscode-test';

async function main() {
  try {
    // The folder containing the Extension Manifest package.json
    // Passed to `--extensionDevelopmentPath`
    const extensionDevelopmentPath = path.resolve(__dirname, '../../');

    // The path to test runner
    // Passed to --extensionTestsPath
    const extensionTestsPath = path.resolve(__dirname, './suite/index');

    // Download VS Code, unzip it and run the integration test
    await runTests({extensionDevelopmentPath, extensionTestsPath});
  } catch (err) {
    console.error('Failed to run tests');
    process.exit(1);
  }
}

main();


src/test/suite/index.ts
// Copyright 2021 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     https://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import * as glob from 'glob';
import * as Mocha from 'mocha';
import * as path from 'path';

export function run(): Promise<void> {
  // Create the mocha test
  const mocha = new Mocha({ui: 'tdd', color: true});

  const testsRoot = path.resolve(__dirname, '..');

  return new Promise((c, e) => {
    glob('**/**.test.js', {cwd: testsRoot}, (err, files) => {
      if (err) {
        return e(err);
      }

      // Add files to the test suite
      files.forEach(f => mocha.addFile(path.resolve(testsRoot, f)));

      try {
        // Run the mocha test
        mocha.run(failures => {
          if (failures > 0) {
            e(new Error(`${failures} tests failed.`));
          } else {
            c();
          }
        });
      } catch (err) {
        console.error(err);
        e(err);
      }
    });
  });
}


src/test/suite/extension.test.ts
// Copyright 2021 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     https://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import * as assert from 'assert';
import * as vscode from 'vscode';

suite('Extension Test Suite', () => {
  vscode.window.showInformationMessage('Start all tests.');
  // TODO: Add zxdb extension tests.
});


src/log.ts
// Copyright 2021 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     https://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import * as vscode from 'vscode';

export const DEBUG =
    vscode.workspace.getConfiguration().get('zxdb.debug') ?? false;
const zxdbOutput = vscode.window.createOutputChannel('zxdb');

function log(type: string, message: string) {
  let logMessage = `[${new Date().toISOString()}] ${type} (zxdb): ${message}`;
  console.log(logMessage);
  zxdbOutput.appendLine(logMessage);
}

export function debug(message: string) {
  if (DEBUG) {
    log('DEBUG', message);
  }
}

export function info(message: string, show: boolean = false) {
  log('INFO', message);
  if (show) {
    vscode.window.showInformationMessage(message);
  }
}

export function warn(message: string) {
  log('WARNING', message);
  vscode.window.showWarningMessage(message);
}

export function error(message: string) {
  log('DEBUG', message);
  vscode.window.showErrorMessage(message);
}


src/console.ts
// Copyright 2021 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     https://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import * as Net from 'net';
import * as vscode from 'vscode';
import * as log from './log';


export const DEFAULT_ZXDB_COMMAND = 'ffx debug connect -- --enable-debug-adapter';
export const DEFAULT_ZXDB_TIMEOUT =
    5000;  // 5 seconds. Usually "ffx debug connect" takes less than 1 second.
export const DEFAULT_SERVER_PORT = 15678;
type TimerCallback = () => void;

export class ZxdbConsole {
  terminal: vscode.Terminal|undefined;
  terminalPID: number|undefined;
  timer: NodeJS.Timeout|undefined;
  timerCallback: TimerCallback|undefined;

  public createZxdbTerminal() {
    if (this.terminal) {
      return;
    }
    try {
      this.terminal = vscode.window.createTerminal(`zxdb console`);
      let command: string =
          vscode.workspace.getConfiguration().get('zxdb.command') ??
          DEFAULT_ZXDB_COMMAND;
      this.terminal.sendText(command);
      this.terminal.show(true);
      this.terminal.processId.then((id) => {
        this.terminalPID = id;
        if (this.terminal) {
          log.info(
              `${this.terminal.name} launched with pid ${this.terminalPID}`);
        }
      });
    } catch (Error) {
      vscode.window.showErrorMessage(Error);
    }
  }

  private sleep(ms: number) {
    return new Promise((resolve) => {
      setTimeout(resolve, ms);
    });
  }


  public async waitForZxdbConsole(timeoutOverride?: number) {
    return new Promise<void>(async (resolve, reject) => {
      // This method waits for the spawned zxdb console to be ready.
      // We check this by connecting to the debug adapter server port.
      // Once the connection succeeds, it means the server is ready.
      // We disconnect and continue debugging.
      let timeout: number =
          vscode.workspace.getConfiguration().get('zxdb.timeout') ??
          DEFAULT_ZXDB_TIMEOUT;
      if (timeoutOverride) {
        timeout = timeoutOverride;
      }
      let socket: Net.Socket;
      let connected = false;
      let retry = true;
      let connect = () => {
        socket = new Net.Socket();
        socket.connect(DEFAULT_SERVER_PORT, 'localhost');

        socket.on('connect', () => {
          connected = true;
          log.info('zxdb console has started.');
          socket.destroy();
          resolve();
        });

        let retryConnect = async () => {
          if (retry) {
            socket.destroy();
            await this.sleep(1000);
            connect();
          }
        };

        socket.on('error', (err) => {
          log.debug('socket error - ' + err);
          retryConnect();
        });
      };


      connect();
      this.timerCallback = () => {
        retry = false;
        if (socket) {
          socket.destroy();
        }
        if (!connected) {
          this.destroyZxdbTerminal();
          log.info('Timeout starting zxdb console.');
          reject(
              'Timeout starting zxdb console. See https://tinyurl.com/zxdb-troubleshooting');
        }
      };

      this.timer = setTimeout(this.timerCallback, timeout);
    });
  }

  public destroyZxdbTerminal() {
    log.info('zxdb console is destroyed.');
    if (this.terminalPID) {
      process.kill(-this.terminalPID);
      this.terminalPID = undefined;
    }
    if (this.terminal) {
      this.terminal.dispose();
      this.terminal = undefined;
    }
    if (this.timer) {
      clearTimeout(this.timer);
      this.timer = undefined;
    }
  }

  public onDidCloseTerminal(t: vscode.Terminal) {
    if (t === this.terminal) {
      if (this.timer) {
        clearTimeout(this.timer);
      }
      if (this.timerCallback) {
        this.timerCallback();
      }
    }
  }
}


src/extension.ts
// Copyright 2021 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     https://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import * as Net from 'net';
import * as vscode from 'vscode';
import {CancellationToken, DebugAdapterTracker, DebugAdapterTrackerFactory, DebugConfiguration, ProviderResult, WorkspaceFolder,} from 'vscode';
import {DebugProtocol} from 'vscode-debugprotocol';

import {DEFAULT_SERVER_PORT, ZxdbConsole} from './console';
import * as log from './log';

let console: ZxdbConsole;

export function activate(context: vscode.ExtensionContext) {
  log.info('zxdb extension is now active!');

  const deprecationNotice = new vscode.MarkdownString('zxdb extension is deprecated. Please use the [Fuchsia extension](https://marketplace.visualstudio.com/items?itemName=fuchsia-authors.vscode-fuchsia).');
  log.warn(deprecationNotice.value);

  console = new ZxdbConsole();
  context.subscriptions.push(
      vscode.commands.registerCommand('extension.zxdb.TestCommand', () => {
        log.info('Starting zxdb.command test', true);
        console.createZxdbTerminal();
        console.waitForZxdbConsole(10 * 60 * 1000)
            .then(
                value => {
                  log.info('zxdb.command test successful!', true);
                  console.destroyZxdbTerminal();
                },
                reason => {
                  log.error('zxdb.command test failed');
                });
      }));

  context.subscriptions.push(vscode.commands.registerCommand(
      'extension.zxdb.pickProcess', (config) => {
        let promptText: string = '';
        if (config.request === 'launch') {
          promptText =
              'Launch(zxdb): Enter name of the process that will be launched.' +
              ' Hint: For components, it\'s usually the component name.';
        } else {
          promptText = 'Attach(zxdb): Enter name of the process to debug.' +
              ' The process should already be running or should be started separately.';
        }
        // In future, add code to run ps and give a list of process to pick
        // from.
        return vscode.window.showInputBox({
          placeHolder: '(e.g. hello-world-test)',
          value: '',
          prompt: promptText
        });
      }));

  context.subscriptions.push(vscode.commands.registerCommand(
      'extension.zxdb.enterLaunchCommand', (config) => {
        return vscode.window.showInputBox({
          placeHolder: '(e.g. fx test hello-world-test)',
          value: '',
          prompt:
              'Launch(zxdb): Enter launch command. This will be run in the vscode terminal.'
        });
      }));


  vscode.window.onDidCloseTerminal((t) => {
    console.onDidCloseTerminal(t);
  });

  // register a configuration provider for 'zxdb' debug type
  const provider = new ZxdbConfigurationProvider();
  context.subscriptions.push(
      vscode.debug.registerDebugConfigurationProvider('zxdb', provider));

  let factory = new ZxdbDebugAdapterFactory();
  context.subscriptions.push(vscode.Disposable.from(
      vscode.debug.registerDebugAdapterDescriptorFactory('zxdb', factory)));

  let logger = new ZxdbDebugAdapterTrackerFactory();
  context.subscriptions.push(vscode.Disposable.from(
      vscode.debug.registerDebugAdapterTrackerFactory('zxdb', logger)));
}

// This method is called when extension is deactivated.
export function deactivate() {}

class ZxdbConfigurationProvider implements vscode.DebugConfigurationProvider {
  maxFilterNameLength = 31;
  // This method is called just before launching debug session.
  // Final updates to debug configuration can be done here.
  resolveDebugConfigurationWithSubstitutedVariables(
      folder: WorkspaceFolder|undefined, config: DebugConfiguration,
      token?: CancellationToken): ProviderResult<DebugConfiguration> {
    // if launch.json is missing or empty
    if (!config.type && !config.request && !config.name) {
      log.error(
          'launch.json does not have any zxdb configuration. Initial configurations will be ' +
          'added automatically, if not add it manually (Hint: Add Configuration -> zxdb...).\n' +
          'Next, pick a configuration in the launch configuration dropdown to start debugging.');
      return null;
    }

    if (config.process.length > this.maxFilterNameLength) {
      config.process = config.process.substring(0, this.maxFilterNameLength);
      log.warn(
          'Process name too long. It will be trimmed to "' + config.process +
          '"');
    }

    if (config.request === 'attach') {
      log.info('Attaching to ' + config.process);
    } else if (config.request === 'launch') {
      log.info(
          'Launching process ' + config.process + ' with command "' +
          config.launchCommand + '"');
    } else {
      log.error('Unknown launch config');
      return null;
    }

    return config;
  }
}
class ZxdbDebugAdapterFactory implements vscode.DebugAdapterDescriptorFactory {
  createDebugAdapterDescriptor(_session: vscode.DebugSession):
      ProviderResult<vscode.DebugAdapterDescriptor> {
    // Start zxdb console and wait for it to be ready.
    // Then create a client to start debugging in vscode.
    console.createZxdbTerminal();
    return console.waitForZxdbConsole().then(() => {
      log.info('Creating debug adapter client.');
      let client = new vscode.DebugAdapterServer(DEFAULT_SERVER_PORT);
      return client;
    });
  }
}

class ZxdbDebugAdapterTracker implements DebugAdapterTracker {
  ignoreError: boolean = false;
  restart: boolean = false;
  launchPID: number|undefined = undefined;
  constructor(private readonly session: vscode.DebugSession) {}

  private toString(data: DebugProtocol.Message|Error|DebugConfiguration) {
    return JSON.stringify(data, null, 2);
  }

  public onWillStartSession() {
    this.ignoreError = false;
    log.info(`Starting debug session:\n${
        this.toString(this.session.configuration)}\n`);
  }

  public onWillReceiveMessage(message: any) {
    let request: DebugProtocol.Request = message;
    if (request.command === 'disconnect') {
      // Ignore errors reported after this. The connection is closed by the
      // backend and hence debug adapter reports connection loss errors which
      // needs to be ignored. It might confuse users if these errors are shown.
      this.ignoreError = true;
      if (request.arguments && request.arguments.restart) {
        this.restart = request.arguments.restart;
      }
    } else if (request.command === 'runInTerminal') {
      let response: DebugProtocol.RunInTerminalResponse = message;
      if (response.body.shellProcessId) {
        this.launchPID = getChildPID(response.body.shellProcessId);
      } else if (response.body.processId) {
        this.launchPID = response.body.processId;
      }
    }
    log.debug(`Sent:\n${this.toString(message)}\n`);
  }

  public onDidSendMessage(message: DebugProtocol.Message) {
    log.debug(`Received:\n${this.toString(message)}\n`);
  }

  public onWillStopSession() {
    log.info('Stopping debug Session.\n');
    if (!this.restart) {
      console.destroyZxdbTerminal();
    }
    destroyZxdbLaunch(this.launchPID);
    this.launchPID = undefined;
  }

  public onError(error: Error) {
    let logFunction = log.error;

    // If error should not be shown to user, log it as info to the output.
    if (this.ignoreError) {
      logFunction = log.info;
    }

    if (error.stack) {
      logFunction(`zxdb ${error.stack}\n`);
    } else {
      logFunction(`zxdb ${error.name}:${error.message}\n`);
    }
  }

  public onExit(code: number|undefined, signal: string|undefined) {
    log.info(`Debug adapter exited! Exit code: ${code}, signal: ${signal}.\n`);
  }
}

export class ZxdbDebugAdapterTrackerFactory implements
    DebugAdapterTrackerFactory {
  public createDebugAdapterTracker(session: vscode.DebugSession):
      ProviderResult<DebugAdapterTracker> {
    return new ZxdbDebugAdapterTracker(session);
  }
}

// Methods to control launched process
function getChildPID(pid: number): number|undefined {
  let childPID = undefined;
  if (process.platform !== 'win32') {
    childPID =
        Number(require('child_process').execSync(`pgrep -P ${pid}`) + '');
    log.debug(`Child process ID: ${childPID}`);
  }
  return childPID;
}

function destroyZxdbLaunch(pid: number|undefined) {
  if (pid) {
    log.info(`zxdb launch (PID:${pid}) is destroyed.`);
    process.kill(-pid, 2);  // SIGINT
  }
}
